### PR TITLE
Use interface for Vault instance

### PR DESCRIPTION
**Pull request by lpichler**
This change also add mock struct which satisfy interface `LogicalClient` as well as `vault.Logical` - it can be used for testing for Vault-related code.

### Links
Pull out from https://github.com/RedHatInsights/sources-api-go/pull/76

### CODE REVIEW

File: dao/authentication_dao.go [hunk 1]
```diff
@@ -210,7 +210,7 @@ func (a *AuthenticationDaoImpl) Create(auth *m.Authentication) error {
 		return err
 	}
 
-	out, err := Vault.Write(path, data)
+	out, err := (*Vault).Write(path, data)
```
Comment (lindgrenj6) [inline dao/authentication_dao.go]: So I'm actually kind of mad at myself that I missed this the first time in the availability status listener pr :disappointed:. 

We shouldn't be using pointers of interfaces because basically an interface is a pointer to a struct that satisfies a "signature", so basically an interface type is _already_ a pointer.

For example see here how I instantiated an Rbac interface var: https://github.com/RedHatInsights/sources-api-go/blob/e5ad9bbed6af58bf92fb9540cefebac7fd42069c/middleware/authorization.go#L20

The Interface var is assigned a pointer to a struct that "satisfies" the interface. 

I noticed it last time and thought a bit about it here: https://github.com/RedHatInsights/sources-api-go/blob/e5ad9bbed6af58bf92fb9540cefebac7fd42069c/dao/common.go#L9 I didn't notice that the `EventModelDao` is actually an interface - therefore we shouldn't be returning pointers to it. Basically a pointer to an interface makes it take double de-referencing which doubles the potential for null-pointer errors. This way we won't have to de-ref the pointer before calling methods on the structs as well. 

also: you can see here https://github.com/RedHatInsights/sources-api-go/blob/e5ad9bbed6af58bf92fb9540cefebac7fd42069c/dao/common.go#L13

that basically we're returning a pointer to a pointer. Should have jumped out to me more but I glossed over it the first time.

---

tl; dr: can we refactor these functions/where they're used to not be pointers to interfaces. Everywhere there is `*EventModelDao` needs to just be `EventModelDao`. 

If this doesn't make sense let me know and I can try and explain it better in person, also if you don't want do do the `EventModelDao` changes in this PR thats fine - we can do them in a follow-up.

File: dao/db.go [hunk 2]
```diff
@@ -16,11 +16,15 @@ var (
 	DB *gorm.DB
 
 	vaultClient *vault.Client
-	Vault       *vault.Logical
+	Vault       *LogicalClient
```
Comment (lindgrenj6) [inline dao/db.go]: This is an interface so shouldn't be a pointer like I said earlier - this is what made me realize I missed it last time.

File: dao/db.go [hunk 3]
```diff
@@ -55,7 +59,8 @@ func Init() {
 	if err != nil {
 		panic(fmt.Sprintf("Failed to Create Vault Client: %v", err))
 	}
-	Vault = vaultClient.Logical()
+
+	SetVault(vaultClient.Logical())
```
Comment (lindgrenj6) [inline dao/db.go]: I actually don't think we need the function here - one can just do:
```go
Vault = &vaultClient.Logical()
````
and in tests, just overwrite it in the same way.

File: dao/interfaces.go [hunk 4]
```diff
@@ -90,3 +91,10 @@ type SourceTypeDao interface {
 	Update(src *m.SourceType) error
 	Delete(id *int64) error
 }
+
+type LogicalClient interface {
```
Comment (lindgrenj6) [inline dao/interfaces.go]: Lets rename this to `VaultClient`.

**Review by lindgrenj6 — CHANGES_REQUESTED** [review]

**Review by MikelAlejoBR — APPROVED** [review]

File: dao/authentication_dao.go (hunk 1 above)
Comment (lpichler) [reply dao/authentication_dao.go]: oh wow that's interesting - it definitely needs to be changed.

Regard to:
https://github.com/RedHatInsights/sources-api-go/blob/e5ad9bbed6af58bf92fb9540cefebac7fd42069c/dao/common.go#L13

It looks same like this 

https://github.com/RedHatInsights/sources-api-go/blob/e5ad9bbed6af58bf92fb9540cefebac7fd42069c/middleware/authorization.go#L20

 I would say it is initialising pointer with address(`&SourceDaoImpl{}`) which seems OK  as `resource` is not pointer on interface but definitely returning values needs to be changed  in the functions and also other occurrence.

I will change `*EventModelDao` in follow up PR and I prefer to do it after series of PRs about adding auths and appauths to listener.

@lindgrenj6 It does make sense and thanks for explanation!

File: dao/db.go (hunk 3 above)
Comment (lpichler) [reply dao/db.go]: when `Vault` is interface variable then just
`Vault = vaultClient.Logical()`

**Comment by lpichler** [comment]
thanks @MikelAlejoBR for review 👍 

and @lindgrenj6 I updated PR according to your suggestions and now it is very little change

File: dao/authentication_dao.go (hunk 1 above)
Comment (lindgrenj6) [reply dao/authentication_dao.go]: Yep that sounds good to me. Just the return type being a pointer is the part I had missed last time - it's easy to add many layers of indirection!

**Review by lindgrenj6 — APPROVED** [review]
